java/com/facebook/soloader/SoLoader.java

```diff
@@ -260,11 +263,27 @@ private static void initSoSources(Context context, int flags, @Nullable SoFileLo
             }
 
             if ((sFlags & SOLOADER_DISABLE_BACKUP_SOSOURCE) != 0) {
-              sBackupSoSource = null;
+              sBackupSoSources = null;
             } else {
-              sBackupSoSource = new ApkSoSource(context, SO_STORE_NAME_MAIN, apkSoSourceFlags);
-              Log.d(TAG, "adding backup  source: " + sBackupSoSource.toString());
-              soSources.add(0, sBackupSoSource);
+
+              String mainApkDir = context.getApplicationInfo().sourceDir;
+              ArrayList<UnpackingSoSource> backupSources = new ArrayList<>();
+              ApkSoSource mainApkSource = new ApkSoSource(context, mainApkDir, SO_STORE_NAME_MAIN, apkSoSourceFlags);
+              backupSources.add(mainApkSource);
+              Log.d(TAG, "adding backup source from : " + mainApkSource.toString());
+
+              if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+                Log.d(TAG, "adding backup sources from split apks");
+                int splitIndex = 0;
+                for (String splitApkDir : context.getApplicationInfo().splitSourceDirs) {
```

Comment:
splitSourceDirs may be null in some cases so I suggest to add a necessary check here.
So does aosp in its sources. e.g https://github.com/aosp-mirror/platform_frameworks_base/blob/master/core/java/android/app/LoadedApk.java#L955